[`flake8-simplify`] Make `SIM114` autofix unsafe when branches have different comments

phongddo opened the pull request:
<!--
Thank you for contributing to Ruff/ty! To help us out with reviewing, please consider the following:

- Does this pull request include a summary of the change? (See below.)
- Does this pull request include a descriptive title? (Please prefix with `[ty]` for ty pull
  requests.)
- Does this pull request include references to any relevant issues?
-->

## Summary

Fixes https://github.com/astral-sh/ruff/issues/19576

Improves `SIM114` rule to handle branch merging with comment preservation safely, offer unsafe fixes when branches contain different comments to prevent comment loss.

## Problem

The original `SIM114` autofix has an issue where it wasn't detecting trailing comments because it was checking comments within `body_range` starting from the **end of line** rather than the **end of test** as documented. 

For example, this case was incorrectly handled:

```python
if x == 1: # 1
    print("Hello")
elif x == 2: # 2
    print("Hello")
```

Running with `--fix` would produce:

```bash
-if x == 1: # 1
-    print("Hello")
-elif x == 2: # 2
+if x == 1 or x == 2: # 1
     print("Hello")

Would fix 1 error.
```

It offers a **safe** fix even though the comments are not identical. I believe this is exactly the case mentioned in the reported issue, so in this situation we should offer unsafe fixes instead. In my opinion, this is a good strategy and consistent with how other rules are handled as @ntBre mentioned in other PR linked to the issue (please let me know what you think). 

## Solution

- Fixed `body_range` to start from the end of the test (as it's documented) instead of the end of the test's line, allowing detection of trailing comments and validation against comments in the other branch before merging branches.  
- Modified `merge_branches` to offer an unsafe fix when the code block is identical but comments differ, instead of skipping the merge entirely. Otherwise, a safe fix is offered as is.
- Updated rule documentation to include `Fix Safety` section to explain unsafe fix in this scenario.

## Manual Verification

✏️ Those ecosystem reports are about unsafe fixes that we offer when merging branches with different comments.

**The above test case → unsafe fix**

```bash
SIM114 Combine `if` branches using logical `or` operator
  |
1 | / if x == 1: # 1
2 | |     print("Hello")
3 | | elif x == 2: # 2
4 | |     print("Hello")
  | |__________________^
  |
help: Combine `if` branches

Found 1 error.
No fixes available (1 hidden fix can be enabled with the `--unsafe-fixes` option).
```

**The issue's test case → unsafe fix** 

```bash
SIM114 Combine `if` branches using logical `or` operator
   |
 1 | / if isinstance(exc, HTTPError) and (
 2 | |     (500 <= exc.status <= 599)
 3 | |     or exc.status == 408  # server errors
 4 | |     or exc.status == 429  # request timeout
 5 | |     ):  # too many requests
 6 | |         return True
 7 | |
 8 | | # Consider all SSL errors as temporary. There are a lot of bug
 9 | | # reports from people where various SSL errors cause a crash
10 | | # but are actually just temporary. On the other hand, we have
11 | | # no information if this ever revealed a problem where retrying
12 | | # was not the right choice.
13 | | elif isinstance(exc, ssl.SSLError):
14 | |     return True
   | |_______________^
   |
help: Combine `if` branches

Found 1 error.
No fixes available (1 hidden fix can be enabled with the `--unsafe-fixes` option).
```

**No comments test case → safe fix** 

```bash
SIM114 [*] Combine `if` branches using logical `or` operator
  |
1 | / if x == 1:
2 | |     print("Hello")
3 | | elif x == 2:
4 | |     print("Hello")
  | |__________________^
  |
help: Combine `if` branches

Found 1 error.
[*] 1 fixable with the `--fix` option.
```

## Test Plan
- Added a comment in snapshot test. I'm not sure how we can distinguish between a safe and an unsafe fix in the snapshot test, so please guide me 🙏

phongddo commented:
Hey @ntBre 👋 I've got this PR up with a suggestion to address the issue https://github.com/astral-sh/ruff/issues/19576. Please take a look whenever you have a moment. Thank you 😄

ntBre commented on crates/ruff_linter/src/rules/flake8_simplify/snapshots/ruff_linter__rules__flake8_simplify__tests__SIM114_SIM114.py.snap:
```diff
@@ -142,329 +117,364 @@ help: Combine `if` branches
+142 | 
+143 | 
+144 | if(x > 200): pass
+note: This is an unsafe fix and may change runtime behavior
```
This added `note` is how we can tell that the fix is now unsafe 👍

ntBre commented on crates/ruff_linter/resources/test/fixtures/flake8_simplify/SIM114.py:
```diff
@@ -4,11 +4,6 @@
-    b
-elif c:  # but not on the second branch
-    b
-
```
Ah this looks like a duplicate test case? I think I'd still lean toward not removing it here because it causes all the line numbers in the snapshot to churn. Same with the new comment below.

phongddo commented:
I reverted the unsafe fix and kept only the `body_range` fix [here](https://github.com/astral-sh/ruff/pull/22079/changes#diff-11a16368fae4b28fc3b47729a9db05e583106d078120ecbf523de5ab43de6773R176). This change would skip the diagnostic and fix entirely where branch comments differ. I also added a test case similar to the one mentioned in the issue (and adjust comments in other test cases). Please let me know what you think 😄

ntBre commented:
Thank you!

Oh no, I think I see why end-of-line comments might have been treated separately :grimacing: There's a newly removed RUF100 unused noqa comment diagnostic in the [ecosystem report](https://github.com/pypa/cibuildwheel/blob/99059b74429eb968569914f1951bdeca074fb7ed/test/utils.py#L263-L277):

```py
if platform == "pyodide" and python_abi_tags is None:
    python_abi_tags = [
        "cp312-cp312",
        "cp313-cp313",
    ]
elif platform == "android" and python_abi_tags is None:  # noqa: SIM114
    python_abi_tags = [
        "cp313-cp313",
        "cp314-cp314",
    ]
elif platform == "ios" and python_abi_tags is None:
    python_abi_tags = [
        "cp313-cp313",
        "cp314-cp314",
    ]
```

I think this comment is now preventing SIM114 from firing, making it "unused" as a noqa comment, but removing it would make SIM114 trigger again. Apparently this has come up before, but I didn't realize it (https://github.com/astral-sh/ruff/issues/6025). Let's make sure to add a comment about this somewhere, it's a bit surprising.

I think we can fix this if we keep the diagnostic for removing end-of-line comments but make the fix unsafe, but suppress the diagnostic for comments like the ones in the issue. Is that doable?

Sorry for the back and forth, but I keep noticing new things in the ecosystem check!

MichaReiser commented:
> I think we can fix this if we keep the diagnostic for removing end-of-line comments but make the fix unsafe, but suppress the diagnostic for comments like the ones in the issue. Is that doable?

Could we detect if any of the comments are a pragma comment instead and not emit a diagnostic in that case?
